### ISSUE TITLE
Make it easy (and documented) how to get debugging info out of dataset definitions

### ISSUE THREAD

**Issue by evansd** [opensafely-core/ehrql#2149]
Because of the way dataset definitions are executed in a separate process and the resulting query model serialized to stdout it's not immediately obvious how to get debugging information out of them e.g. to check that command line arguments you have supplied are being parsed correctly.

Here's an example of a user asking for exactly this:
https://bennettoxford.slack.com/archives/C01D7H9LYKB/p1728574390671559?thread_ts=1728573740.085229&cid=C01D7H9LYKB

Because we pass through stderr it is _possible_ to get debug information out:
https://github.com/opensafely-core/ehrql/blob/971924e43dee8ea8b48c501fe124c49a688a1a1d/ehrql/loaders.py#L85-L89

But you either need to know how to write to stderr:
```py
print("debug stuff", file=sys.stderr)
```

Or rely on the fact that ehrQL has already configured logging by the time your dataset definitions is loaded:
```py
import logging

logging.info("debug stuff")
```

That results in output which looks like (note `debug stuff` on line 2):
```
2024-10-10 16:19:47 [info     ] Compiling dataset definition from demo_dataset.py
2024-10-10 16:19:48 [info     ] debug stuff [root]
2024-10-10 16:19:48 [info     ] Generating dummy dataset
2024-10-10 16:19:48 [info     ] Attempting to generate 10 matching patients (random seed: BwRV3spP, timeout: 60s)
2024-10-10 16:19:48 [info     ] Use `dataset.configure_dummy_data(population_size=N)` to change number of patients generated
2024-10-10 16:19:48 [info     ] Building dataset and writing results
```

Neither of these are particular great though.

I think it would be much better if we could support plain old `print()` statements as a way to get debugging information out (at the moment, doing this would result in mysterious deserialization errors).

One option would be to use a named pipe or something to get the serialized query model out of the subprocess and forward through both `stdout` and `stderr` unmodified. But I think that's unnecessarily complicated and we can just read the serialized query model off the _end_ of stdout and then pass through anything that comes before as user-generated output to stderr. If we ensure our blob of JSON always starts `\n{` then we can reliably identify it as the last instance of that sequence in the output stream.

**Comment by bloodearnest** [comment]
Ah, podman gets it right, as it's not client/server like docker: https://docs.podman.io/en/latest/markdown/podman-exec.1.html

https://docs.podman.io/en/latest/markdown/podman-exec.1.html

Damn, that's a useful thing to be able to do with a stateless ephemeral container

**Comment by evansd** [comment]
Oh, the file descriptor thing might work fine actually. This is about communication across the pledge isolation boundary _within_ ehrQL (so all running within the container) not across the docker/client boundary.

**Comment by bloodearnest** [comment]
Ah! gotcha. Sweet.

**Comment by evansd** [comment]
Further discussion in Slack thread:
https://bennettoxford.slack.com/archives/C069YDR4NCA/p1733135371374899

**Comment by evansd** [comment]
Closed by #2275
